torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -173,22 +173,27 @@ PyObject* functionToPyObject(std::shared_ptr<Function> cdata)
     return obj;
   }
 
-  auto& fn = *cdata;
-  auto it = cpp_function_types.find(std::type_index(typeid(fn)));
-  if (it == cpp_function_types.end()) {
-    return PyErr_Format(PyExc_TypeError,
-        "Don't know how to create Python object for %s", typeid(fn).name());
-  }
+  if (cdata->pyobj) {
+    Py_INCREF(cdata->pyobj);
+  } else {
+    auto& fn = *cdata;
+    auto it = cpp_function_types.find(std::type_index(typeid(fn)));
+    if (it == cpp_function_types.end()) {
+      return PyErr_Format(PyExc_TypeError,
+          "Don't know how to create Python object for %s", typeid(fn).name());
+    }
 
-  PyTypeObject* type = (PyTypeObject*)it->second.get();
-  THPObjectPtr obj(type->tp_alloc(type, 0));
-  if (!obj) return NULL;
-  THPCppFunction* f = (THPCppFunction*)obj.get();
-  new (&f->cdata) std::shared_ptr<Function>(cdata);
-  if (!f->cdata) {
-    return NULL;
+    PyTypeObject* type = (PyTypeObject*)it->second.get();
+    THPObjectPtr obj(type->tp_alloc(type, 0));
+    if (!obj) return NULL;
+    THPCppFunction* f = (THPCppFunction*)obj.get();
+    f->cdata = cdata;
+
+    cdata->pyobj = obj.release();
+    Py_INCREF(cdata->pyobj);
```

Comment:
Ho my bad, the refcount should be 1 indeed because we have a weak reference.
But the thing is that I have just reused the logic used for variables [here](https://github.com/pytorch/pytorch/blob/3c275fe7a08411e23557107d00b358c304148e75/torch/csrc/autograd/python_variable.cpp#L40-L41).